lib/puppet/type/certificate_file.rb

```diff
@@ -0,0 +1,57 @@
+Puppet::Type.newtype(:certificate_file) do
+  desc 'Manages an x509 certificate file from a source certificate'
+
+  ensurable
+
+  def exists?
+    self[:ensure] == :present
+  end
+
+  newparam(:path, :namevar => true) do
+    desc "Path the certificate will be copied to."
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the truststore. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the truststore. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the truststore. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  newproperty(:source_cert) do
+    desc "Specifies the source certificate that will be copied to the declared location."
+  end
+
+  def generate
+    file_opts = {
+      ensure: (self[:ensure] == :absent) ? :absent : :file,
+      source: self[:source_cert],
```

Comment:
The source_cert has to be implemented as a property and validated
inside the provider code. The validate block of a type executes during
catalog compilation and does not give access to the certificate on disk. 
Thus we pass it to the source attribute on the file to allow the File to 
handle the source.